server/src/main/java/org/opensearch/search/profile/query/ConcurrentQueryProfileBreakdown.java

```diff
@@ -44,14 +63,262 @@ public AbstractProfileBreakdown<QueryTimingType> context(Object context) {
+                // compute max/min/avg TimingType time across slices
+                queryBreakdownMap.compute(
+                    maxBreakdownTypeTime,
+                    (key, value) -> (value == null) ? sliceBreakdownTypeTime : Math.max(sliceBreakdownTypeTime, value)
+                );
+                queryBreakdownMap.compute(
+                    minBreakdownTypeTime,
+                    (key, value) -> (value == null) ? sliceBreakdownTypeTime : Math.min(sliceBreakdownTypeTime, value)
+                );
+                queryBreakdownMap.compute(
+                    avgBreakdownTypeTime,
+                    (key, value) -> (value == null) ? sliceBreakdownTypeTime : sliceBreakdownTypeTime + value
+                );
+
+                // compute max/min/avg TimingType count across slices
+                queryBreakdownMap.compute(
+                    maxBreakdownTypeCount,
+                    (key, value) -> (value == null) ? sliceBreakdownTypeCount : Math.max(sliceBreakdownTypeCount, value)
+                );
+                queryBreakdownMap.compute(
+                    minBreakdownTypeCount,
+                    (key, value) -> (value == null) ? sliceBreakdownTypeCount : Math.min(sliceBreakdownTypeCount, value)
+                );
+                queryBreakdownMap.compute(
+                    avgBreakdownTypeCount,
+                    (key, value) -> (value == null) ? sliceBreakdownTypeCount : sliceBreakdownTypeCount + value
+                );
+
+                // query start/end time for a TimingType is min/max of start/end time across slices for that TimingType
+                queryTimingTypeEndTime = Math.max(queryTimingTypeEndTime, sliceBreakdown.getValue().get(sliceEndTimeForTimingType));
+                queryTimingTypeStartTime = Math.min(queryTimingTypeStartTime, sliceBreakdown.getValue().get(sliceStartTimeForTimingType));
+                queryTimingTypeCount += sliceBreakdownTypeCount;
             }
+            queryBreakdownMap.put(timingTypeKey, queryTimingTypeEndTime - queryTimingTypeStartTime);
+            queryBreakdownMap.put(timingTypeCountKey, queryTimingTypeCount);
+            queryBreakdownMap.compute(avgBreakdownTypeTime, (key, value) -> (value == null) ? 0 : value / sliceLevelBreakdowns.size());
+            queryBreakdownMap.compute(avgBreakdownTypeCount, (key, value) -> (value == null) ? 0 : value / sliceLevelBreakdowns.size());
+            // compute query end time using max of query end time across all timing types
+            queryEndTime = Math.max(queryEndTime, queryTimingTypeEndTime);
         }
+        queryNodeTime = queryEndTime - createWeightStartTime;
+        return queryBreakdownMap;
+    }
+
+    @Override
+    public long toNodeTime() {
+        return queryNodeTime;
+    }
+
+    @Override
+    public void associateCollectorToLeaves(Collector collector, LeafReaderContext leaf) {
+        sliceCollectorToLeaves.computeIfAbsent(collector, k -> new ArrayList<>()).add(leaf);
```

Comment:
Referring back to 

> I believe the profile is shared between index searcher threads, right?

If that is the case, we have data race here for updating the list, please use `compute` method to guarantee atomic  updates:

```
sliceCollectorToLeaves.compute(collector, k, v -> ...);
```

Reply:
@reta good catch. We will need to make `sliceCollectorToLeaves` concurrentHashMap. The list update for a single key itself will be done by single slice thread always.

Reply:
@sohami we better off go just one (small) step ahead and cover that (safety first)

Reply:
But that is the invariant which is assumed in multiple places (like TopNCollector having separate non-concurrent priority queues per slice). So I don't see any risk for this. I will add a comment to explain why synchronization for list value is not needed. wdyt ?

Reply:
> I will add a comment to explain why synchronization for list value is not needed. wdyt ?

:+1: , thanks @sohami